File: src/test/java/dev/jlo/gamemodes/persistence/SqlitePersistenceIntegrationTest.java
package dev.jlo.gamemodes.persistence;

import dev.jlo.gamemodes.player.LocationSnapshot;
import dev.jlo.gamemodes.player.PendingRestore;
import dev.jlo.gamemodes.player.PendingRestoreRepository;
import dev.jlo.gamemodes.player.PlayerSnapshot;
import dev.jlo.gamemodes.player.RestoreState;
import dev.jlo.gamemodes.reward.Reward;
import dev.jlo.gamemodes.reward.RewardOutbox;
import dev.jlo.gamemodes.reward.RewardState;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SqlitePersistenceIntegrationTest {
    private final Clock clock = Clock.fixed(Instant.parse("2026-08-11T12:00:00Z"), ZoneOffset.UTC);

    @Test
    void rewardLeaseSurvivesCloseAndReopenWithClaimedSchemaState() throws IOException, SQLException {
        Path database = Files.createTempFile("gamemodes-reward", ".sqlite");
        Files.deleteIfExists(database);
        UUID player = UUID.randomUUID();
        try {
            try (var connection = DriverManager.getConnection("jdbc:sqlite:" + database.toAbsolutePath())) {
                RewardOutbox outbox = new RewardOutbox(connection, clock);
                assertTrue(outbox.enqueue("match-1", player, "WIN_TOKEN", 1));
                List<Reward> claimed = outbox.claim(1, 60_000);
                assertEquals(1, claimed.size());
                assertEquals(RewardState.CLAIMED, claimed.getFirst().getState());
            }

            try (var connection = DriverManager.getConnection("jdbc:sqlite:" + database.toAbsolutePath())) {
                RewardOutbox outbox = new RewardOutbox(connection, clock);
                List<Reward> claimed = outbox.claim(1, 60_000);
                assertTrue(claimed.isEmpty(), "an unexpired lease must not be claimed twice");
                String state;
                try (var statement = connection.prepareStatement(
                        "SELECT state FROM reward_outbox WHERE match_id='match-1'")) {
                    try (var rows = statement.executeQuery()) {
                        rows.next();
                        state = rows.getString(1);
                    }
                }
                assertEquals("CLAIMED", state);
            }
        } finally {
            Files.deleteIfExists(database);
        }
    }

    @Test
    void pendingPlayerRestoreSurvivesCloseAndReopen() throws IOException, SQLException {
        Path database = Files.createTempFile("gamemodes-restore", ".sqlite");
        Files.deleteIfExists(database);
        UUID player = UUID.randomUUID();
        PlayerSnapshot snapshot = new PlayerSnapshot(
                player,
                List.of(),
                null,
                List.of(),
                null,
                List.of(),
                Map.of("minecraft:max_health", 20.0),
                17.0,
                16,
                4.0,
                0.25f,
                12,
                "SURVIVAL",
                false,
                false,
                new LocationSnapshot("world", 1.0, 65.0, 2.0, 90f, 0f));
        try {
            try (var connection = DriverManager.getConnection("jdbc:sqlite:" + database.toAbsolutePath())) {
                new PendingRestoreRepository(connection, clock).put(snapshot);
            }
            try (var connection = DriverManager.getConnection("jdbc:sqlite:" + database.toAbsolutePath())) {
                PendingRestoreRepository repository = new PendingRestoreRepository(connection, clock);
                PendingRestore claimed = repository.claim(player);
                assertNotNull(claimed);
                assertEquals(RestoreState.CLAIMED, claimed.getState());
                assertEquals(snapshot, claimed.getSnapshot());
                assertTrue(repository.markRestored(player));
                assertEquals(RestoreState.RESTORED, repository.get(player).getState());
            }
        } finally {
            Files.deleteIfExists(database);
        }
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/persistence/SqlitePersistenceIntegrationTest.kt
package dev.jlo.gamemodes.persistence

import dev.jlo.gamemodes.player.LocationSnapshot
import dev.jlo.gamemodes.player.PendingRestoreRepository
import dev.jlo.gamemodes.player.PlayerSnapshot
import dev.jlo.gamemodes.player.RestoreState
import dev.jlo.gamemodes.reward.RewardOutbox
import dev.jlo.gamemodes.reward.RewardState
import java.nio.file.Files
import java.sql.DriverManager
import java.time.Clock
import java.time.Instant
import java.time.ZoneOffset
import java.util.UUID
import kotlin.io.path.absolutePathString
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SqlitePersistenceIntegrationTest {
    private val clock = Clock.fixed(Instant.parse("2026-08-11T12:00:00Z"), ZoneOffset.UTC)

    @Test
    fun `reward lease survives close and reopen with CLAIMED schema state`() {
        val database = Files.createTempFile("gamemodes-reward", ".sqlite")
        Files.deleteIfExists(database)
        val player = UUID.randomUUID()
        try {
            DriverManager.getConnection("jdbc:sqlite:${database.absolutePathString()}").use { connection ->
                val outbox = RewardOutbox(connection, clock)
                assertTrue(outbox.enqueue("match-1", player, "WIN_TOKEN", 1))
                val claimed = outbox.claim(1, 60_000)
                assertEquals(1, claimed.size)
                assertEquals(RewardState.CLAIMED, claimed.single().state)
            }

            DriverManager.getConnection("jdbc:sqlite:${database.absolutePathString()}").use { connection ->
                val outbox = RewardOutbox(connection, clock)
                val claimed = outbox.claim(1, 60_000)
                assertTrue(claimed.isEmpty(), "an unexpired lease must not be claimed twice")
                val state = connection.prepareStatement(
                    "SELECT state FROM reward_outbox WHERE match_id='match-1'"
                ).use { statement -> statement.executeQuery().use { rows -> rows.next(); rows.getString(1) } }
                assertEquals("CLAIMED", state)
            }
        } finally {
            Files.deleteIfExists(database)
        }
    }

    @Test
    fun `pending player restore survives close and reopen`() {
        val database = Files.createTempFile("gamemodes-restore", ".sqlite")
        Files.deleteIfExists(database)
        val player = UUID.randomUUID()
        val snapshot = PlayerSnapshot(
            playerId = player,
            inventory = emptyList(),
            cursor = null,
            armor = emptyList(),
            offhand = null,
            effects = emptyList(),
            attributes = mapOf("minecraft:max_health" to 20.0),
            health = 17.0,
            food = 16,
            saturation = 4.0,
            experience = 0.25f,
            level = 12,
            gameMode = "SURVIVAL",
            allowFlight = false,
            flying = false,
            returnLocation = LocationSnapshot("world", 1.0, 65.0, 2.0, 90f, 0f)
        )
        try {
            DriverManager.getConnection("jdbc:sqlite:${database.absolutePathString()}").use { connection ->
                PendingRestoreRepository(connection, clock).put(snapshot)
            }
            DriverManager.getConnection("jdbc:sqlite:${database.absolutePathString()}").use { connection ->
                val repository = PendingRestoreRepository(connection, clock)
                val claimed = assertNotNull(repository.claim(player))
                assertEquals(RestoreState.CLAIMED, claimed.state)
                assertEquals(snapshot, claimed.snapshot)
                assertTrue(repository.markRestored(player))
                assertEquals(RestoreState.RESTORED, repository.get(player)?.state)
            }
        } finally {
            Files.deleteIfExists(database)
        }
    }
}
